In [ ]:
import pandas as pd
autos = pd.read_csv('https://query.data.world/s/uhjmdv345sinwhtacvscflyvokavaw?dws=00000', encoding = "Latin-1")

In [ ]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

Several variables have missing values: vehicleType;  gearbox; fuelType; notRepairedDamage, with less than 20% null values.

- Change the columns from camelcase to snakecase.
- Change a few wordings to more accurately describe the columns.

In [ ]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

Data Exploration

In [ ]:
autos.describe(include = "all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model    odometer_km  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [ ]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   0    benzin  volkswagen               NaN   
1       125000                   5    diesel        audi                ja   
2       125000                   8    diesel        jeep               NaN   
3       150000                   6    benzin  volkswagen              nein   
4        90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

In [ ]:
autos["price"].unique().shape

(5597,)

In [42]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [43]:
autos["price"].value_counts().head(20)

price
0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: count, dtype: int64

There are 10,778 cars with $0 price, which is around 3% of the cars, we might consider removing those rows.

In [53]:
autos["price"].value_counts().sort_index(ascending = False).head(30)

price
2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
1111111        2
1010010        1
1000000        5
999999        13
999990         1
911911         1
849000         1
820000         1
780000         1
745000         2
Name: count, dtype: int64

In [52]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

price
0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: count, dtype: int64

Given that eBay is an auction site, there could legitimately be items where the opening bid is $1. We will keep the $1 item, but remove anything above $745000.

In [54]:
autos = autos[autos["price"].between(1,745000)]

In [55]:
autos.describe()

price  registration_year       power_ps    odometer_km  \
count  360662.000000      360662.000000  360662.000000  360662.000000   
mean     5937.574377        2004.432771     116.538759  125667.674998   
std      9992.653688          81.014108     189.886460   39827.807548   
min         1.000000        1000.000000       0.000000    5000.000000   
25%      1250.000000        1999.000000      72.000000  100000.000000   
50%      3000.000000        2004.000000     105.000000  150000.000000   
75%      7490.000000        2008.000000     150.000000  150000.000000   
max    745000.000000        9999.000000   20000.000000  150000.000000   

       registration_month  num_photos    postal_code  
count       360662.000000    360662.0  360662.000000  
mean             5.795983         0.0   50997.667564  
std              3.684398         0.0   25759.493178  
min              0.000000         0.0    1067.000000  
25%              3.000000         0.0   30823.000000  
50%              6.000000         0.0   49751.000000  
75%              9.000000         0.0   71672.000000  
max             12.000000         0.0   99998.000000

In [56]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [58]:
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False)

date_crawled
2016-04-03    0.038809
2016-04-04    0.037625
2016-03-20    0.036400
2016-03-14    0.036328
2016-03-12    0.036244
2016-03-21    0.035690
2016-03-07    0.035654
2016-03-19    0.035271
2016-04-02    0.035094
2016-03-28    0.035061
2016-04-01    0.034148
2016-03-29    0.034123
2016-03-09    0.034112
2016-03-30    0.033533
2016-03-08    0.033469
2016-03-15    0.033422
2016-03-25    0.032798
2016-03-11    0.032770
2016-03-10    0.032648
2016-03-22    0.032493
2016-03-23    0.032005
2016-03-26    0.031972
2016-03-31    0.031875
2016-03-17    0.031650
2016-03-27    0.030228
2016-03-16    0.030203
2016-03-24    0.029914
2016-03-05    0.025547
2016-03-13    0.015782
2016-03-06    0.014482
2016-03-18    0.013118
2016-04-05    0.012785
2016-04-06    0.003128
2016-04-07    0.001619
Name: proportion, dtype: float64

In [59]:
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

date_crawled
2016-03-05     9214
2016-03-06     5223
2016-03-07    12859
2016-03-08    12071
2016-03-09    12303
2016-03-10    11775
2016-03-11    11819
2016-03-12    13072
2016-03-13     5692
2016-03-14    13102
2016-03-15    12054
2016-03-16    10893
2016-03-17    11415
2016-03-18     4731
2016-03-19    12721
2016-03-20    13128
2016-03-21    12872
2016-03-22    11719
2016-03-23    11543
2016-03-24    10789
2016-03-25    11829
2016-03-26    11531
2016-03-27    10902
2016-03-28    12645
2016-03-29    12307
2016-03-30    12094
2016-03-31    11496
2016-04-01    12316
2016-04-02    12657
2016-04-03    13997
2016-04-04    13570
2016-04-05     4611
2016-04-06     1128
2016-04-07      584
Name: count, dtype: int64

The date crawled is uniformly distributed, scraped from March 2016 to the early April of 2016.

In [61]:
autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False)

last_seen
2016-04-06    0.218947
2016-04-07    0.130435
2016-04-05    0.126964
2016-03-17    0.028703
2016-04-04    0.025534
2016-04-03    0.025317
2016-04-02    0.024965
2016-03-31    0.024242
2016-04-01    0.023895
2016-03-30    0.023723
2016-03-29    0.023282
2016-03-12    0.023243
2016-03-28    0.022187
2016-03-22    0.020507
2016-03-21    0.020030
2016-03-20    0.019888
2016-03-24    0.019162
2016-03-25    0.018998
2016-03-23    0.018020
2016-03-27    0.016719
2016-03-16    0.016417
2016-03-19    0.016331
2016-03-15    0.016323
2016-03-26    0.015957
2016-03-11    0.012954
2016-03-14    0.012175
2016-03-10    0.011459
2016-03-09    0.009824
2016-03-13    0.008410
2016-03-08    0.007938
2016-03-18    0.006887
2016-03-07    0.005202
2016-03-06    0.004098
2016-03-05    0.001264
Name: proportion, dtype: float64

In [64]:
autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index()

last_seen
2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007938
2016-03-09    0.009824
2016-03-10    0.011459
2016-03-11    0.012954
2016-03-12    0.023243
2016-03-13    0.008410
2016-03-14    0.012175
2016-03-15    0.016323
2016-03-16    0.016417
2016-03-17    0.028703
2016-03-18    0.006887
2016-03-19    0.016331
2016-03-20    0.019888
2016-03-21    0.020030
2016-03-22    0.020507
2016-03-23    0.018020
2016-03-24    0.019162
2016-03-25    0.018998
2016-03-26    0.015957
2016-03-27    0.016719
2016-03-28    0.022187
2016-03-29    0.023282
2016-03-30    0.023723
2016-03-31    0.024242
2016-04-01    0.023895
2016-04-02    0.024965
2016-04-03    0.025317
2016-04-04    0.025534
2016-04-05    0.126964
2016-04-06    0.218947
2016-04-07    0.130435
Name: proportion, dtype: float64

In [65]:
autos["registration_year"].describe()

count    360662.000000
mean       2004.432771
std          81.014108
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The last three days have unexpected high last seen items, 6-10x the values from previous days, which might due to ending crawling days, other than sale spikes.

There are also some odd values in the registration year. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

Find the invalid car registration years

In [70]:
(~autos["registration_year"].between(1990,2016)).sum()/autos.shape[0]

0.0658982648574011

Around 6.6% of our data has invalid car registration years and we may remove them.

In [71]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000    0.066694
1999    0.063547
2005    0.062667
2006    0.057703
2001    0.056953
2003    0.056552
2004    0.056175
2002    0.054285
2007    0.050495
1998    0.049687
Name: proportion, dtype: float64

It appears that most of the vehicles were first registered in the past 18 years since the date crawled.

Explore price by brand

In [76]:
autos["brand"].value_counts(normalize = True)

brand
volkswagen        0.211684
bmw               0.109871
opel              0.106401
mercedes_benz     0.096840
audi              0.089539
ford              0.068912
renault           0.047513
peugeot           0.030151
fiat              0.025689
seat              0.018659
skoda             0.015686
mazda             0.015383
smart             0.014330
citroen           0.013949
nissan            0.013597
toyota            0.012931
hyundai           0.009972
sonstige_autos    0.009513
mini              0.009383
volvo             0.009147
mitsubishi        0.008235
honda             0.007531
kia               0.006914
suzuki            0.006363
alfa_romeo        0.006308
porsche           0.006248
chevrolet         0.005022
chrysler          0.003862
dacia             0.002495
jeep              0.002192
land_rover        0.002166
daihatsu          0.002160
subaru            0.002117
jaguar            0.001736
saab              0.001465
daewoo            0.001457
trabant           0.00

As the top four brands constitute over 50% of the car brands in eBay, we choose them as common brands in the auction website.

In [85]:
brand_counts = autos["brand"].value_counts(normalize = True)
common_brands = brand_counts[brand_counts > 0.09].index

autos[autos["brand"].isin(common_brands)]

date_crawled  \
0       2016-03-24 11:52:17   
3       2016-03-17 16:54:04   
5       2016-04-04 17:36:23   
9       2016-03-17 10:53:50   
11      2016-04-07 10:06:22   
...                     ...   
371518  2016-04-02 20:37:03   
371521  2016-03-27 20:36:20   
371525  2016-03-19 18:57:12   
371526  2016-03-20 19:41:08   
371527  2016-03-07 19:39:19   

                                                     name  seller offer_type  \
0                                              Golf_3_1.6  privat    Angebot   
3                                      GOLF_4_1_4__3TÜRER  privat    Angebot   
5       BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...  privat    Angebot   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...  privat    Angebot   
11          Volkswagen_Passat_Variant_2.0_TDI_Comfortline  privat    Angebot   
...                                                   ...     ...        ...   
371518                          Bmw_320_D_DPF_Touring_!!!  privat    Angebot   
371521                 Opel_Zafira_1.6_Elegance_TÜV_12/16  privat    Angebot   
371525                 Volkswagen_Multivan_T4_TDI_7DC_UY2  privat    Angebot   
371526                             VW_Golf_Kombi_1_9l_TDI  privat    Angebot   
371527       BMW_M135i_vollausgestattet_NP_52.720____Euro  privat    Angebot   

        price  ab_test vehicle_type  registration_year  gearbox  power_ps  \
0         480     test          NaN               1993  manuell         0   
3        1500     test   kleinwagen               2001  manuell        75   
5         650     test    limousine               1995  manuell       102   
9         999     test   kleinwagen               1998  manuell       101   
11       2799  control        kombi               2005  manuell       140   
...       ...      ...          ...                ...      ...       ...   
371518   3999     test        kombi               2005  manuell         3   
371521   1150  control          bus               2000  manuell         0   
371525   9200     test          bus               1996  manuell       102   
371526   3400     test        kombi               2002  manuell       100   
371527  28990  control    limousine               2013  manuell       320   

              model  odometer_km  registration_month fuel_type       brand  \
0              golf       150000                   0    benzin  volkswagen   
3              golf       150000                   6    benzin  volkswagen   
5               3er       150000                  10    benzin         bmw   
9              golf       150000                   0       NaN  volkswagen   
11           passat       150000                  12    diesel  volkswagen   
...             ...          ...                 ...       ...         ...   
371518          3er       150000                   5    diesel         bmw   
371521       zafira       150000                   3    benzin        opel   
371525  transporter       150000                   3    diesel  volkswagen   
371526         golf       150000                   6    diesel  volkswagen   
371527      m_reihe        50000                   8    benzin         bmw   

       unrepaired_damage           ad_created  num_photos  postal_code  \
0                    NaN  2016-03-24 00:00:00           0        70435   
3                   nein  2016-03-17 00:00:00           0        91074   
5                     ja  2016-04-04 00:00:00           0        33775   
9                    NaN  2016-03-17 00:00:00           0        27472   
11                    ja  2016-04-07 00:00:00           0        57290   
...                  ...                  ...         ...          ...   
371518              nein  2016-04-02 00:00:00           0        81825   
371521              nein  2016-03-27 00:00:00           0        26624   
371525              nein  2016-03-19 00:00:00           0        87439   
371526               NaN  2016-03-20 00:00:00           0        40764   
37152

In [102]:
common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz'], dtype='object', name='brand')

In [91]:
brand_price = {}
for brand in common_brands:
    avg_price = autos.loc[autos["brand"] == brand, "price"].mean()
    brand_price[brand] = avg_price

In [92]:
brand_price

{'volkswagen': 5400.188763830599,
 'bmw': 8489.620960331837,
 'opel': 2971.8999132509216,
 'mercedes_benz': 8579.526583862033}

Among common brands, mercedes_benz and bmw have highest average price, around $8500, 1.5-3x the average prices of other common brands.

Store aggregate data in a dataframe

In [96]:
bmp_series = pd.Series(brand_price)

In [97]:
df = pd.DataFrame(bmp_series, columns = ['mean_price'])

In [98]:
df

mean_price
volkswagen     5400.188764
bmw            8489.620960
opel           2971.899913
mercedes_benz  8579.526584

In [106]:
brand_mileage = {}
for brand in common_brands: 
    avg_mileage = autos.loc[autos["brand"] == brand, "odometer_km"].mean()
    brand_mileage[brand] = avg_mileage

In [108]:
df["brand_mileage"] = brand_mileage 

In [109]:
df

mean_price  brand_mileage
volkswagen     5400.188764  128386.044040
bmw            8489.620960  132790.422935
opel           2971.899913  128722.619822
mercedes_benz  8579.526584  130566.228815

There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage. However, the average mileage among common brands did not vary much as the average price did, falling within 10% for top brands.